Add support for IP address changes

Erik142 opened the pull request:
This PR improves upon the existing "Wait for IP step", by continuously checking for changes to the IP address for the VM, even after the IP address has been retrieved.

The implementation is using a simple goroutine which encapsulates the existing method used to wait for the IP address. When an IP address has been found, the goroutine will put the IP address directly in the state, which will be queried when the IP address should be retrieved.

A minor customization has been made to the "SSH port forward step", to retrieve the IP address from the state instead of using a string variable.

Closes #61

sjevs commented:
JFYI This Merge Request didn't work for me - same issue
```
TCP connection to SSH ip/port failed: dial tcp 192.168.178.104:22: connect: connection refused
TCP connection to SSH ip/port failed: dial tcp 192.168.178.104:22: i/o timeout
TCP connection to SSH ip/port failed: dial tcp 192.168.178.104:22: connect: host is down
TCP connection to SSH ip/port failed: dial tcp 192.168.178.104:22: connect: host is down
TCP connection to SSH ip/port failed: dial tcp 192.168.178.104:22: connect: host is down
TCP connection to SSH ip/port failed: dial tcp 192.168.178.104:22: i/o timeout
```

Erik142 commented:
Thank you for this feedback @sjevs! The PR has worked flawlessly for me. Could you describe in more detail how to reproduce the fault you described? A minimal set of steps to reproduce the issue would be preferable, that way I can start to work on a fix directly. Or if possible, please post the complete Packer HCL file that you are using when this error occurs.

Thanks!

hoerup commented:
while your'e at it 
ip_getter is not a documented option could you add it to docs/builders/iso/xenserver-iso.html.markdown ?

Erik142 commented:
@hoerup I could definitely take a look at that. However, I believe that should be a separate issue/pull request, no? If you think it's OK to include it in this PR, I can take care of that!

ddelnano commented on builder/xenserver/common/step_wait_for_ip.go:
```diff
@@ -29,47 +29,62 @@ func (self *StepWaitForIP) Run(ctx context.Context, state multistep.StateBag) mu
+		}
+	}(*c, ui, config)
+
+	time.Sleep(500 * time.Millisecond)
```
What's the purpose of this `time.Sleep`?

Erik142 replied:
The purpose of this `time.sleep` is, as you mentioned in your other comment, a minimum wait time to prevent the `Predicate` from resolving early. I have replaced the hardcoded value with a configurable one instead.

ddelnano commented on builder/xenserver/common/step_wait_for_ip.go:
```diff
@@ -29,47 +29,62 @@ func (self *StepWaitForIP) Run(ctx context.Context, state multistep.StateBag) mu
+	err = InterruptibleWait{
+		Timeout:           self.Timeout,
+		PredicateInterval: 5 * time.Second,
+		Predicate: func() (result bool, err error) {
+			var ok bool
+			if ip, ok = state.Get("instance_ssh_address").(string); ok && ip != "" {
+				return true, nil
+			}
```
It seems like the `InterruptibleWait` could still return a stale value. I'm wondering if it makes sense to introduce a minimum wait time to prevent the `Predicate` from resolving early. This would allow end users to further delay resolving the predicate if the default runtime isn't enough time for their DHCP to converge. I think we could account for the existing `time.Sleep` call above as the default for this value. 

In addition to that, it looks like the Timeout is hard coded to 300 seconds for the xva builder ([source](https://github.com/ddelnano/packer-plugin-xenserver/blob/81c1ea03eec8c7bd11e87d80e1b7822d3ab46a00/builder/xenserver/xva/builder.go#L164)). We should update to be consistent with the [iso version](https://github.com/ddelnano/packer-plugin-xenserver/blob/81c1ea03eec8c7bd11e87d80e1b7822d3ab46a00/builder/xenserver/iso/builder.go#L276) so the maximum can be tuned as well.

Erik142 replied:
Good suggestion! I have added a configurable wait time, named "dhcp_wait" that will be used as an initial wait time before executing the `Predicate`. I have also, in a separate commit, aligned the XVA builder so that it uses the same `self.config.InstallTimeout` value as the ISO builder. However, I could not successfully update "config.hcl2spec.go" by executing `make generate`. I got an error that I have not (yet) been able to understand. If you haven't seen it before, I'll spend some time to figure it out, as I assume that I have to include the changes made by `make generate` in this PR? The error I get is the following:

```bash
mapstructure-to-hcl2: common.[Config]: internal error: package "fmt" without types was imported from "github.com/xenserver/packer-builder-xenserver/builder/xenserver/common"
builder/xenserver/common/config.go:1: running "packer-sdc": exit status 1
make: *** [generate] Error 1
```

Erik142 commented:
Hi @ddelnano, apologies for the delay on my side as well. I've been having quite a lot to tend to personally for a while. I have answered and addressed your comments. I have one question regarding `make generate` which I hope you will be able to answer for me! Other than that, please have another look at this PR. If there is anything else you want me to change, just let me know!

enkaskal commented:
ran into this issue as well, fetched @Erik142's branch, rebased to current master and it solved my issue as well.

N.B. I did have to install guest-utilities as part of the d-i preseed in order for it to work; although I believe that is intended based on the original code.

while my testing was done `go build -o ...` and `packer plugins install --path ...` I did notice that the `make generate` error happens on master identical to the branch so not sure this should be a blocker.

@ddelnano if that's the case, what if any hold ups are remaining on this PR?

nathanael-h commented:
Hey @enkaskal thanks for your message. I'll try to reboot and merge this old PR. 

> N.B. I did have to install guest-utilities as part of the d-i preseed in order for it to work; although I believe that is intended based on the original code.

Indeed, the VM IP information goes like this: VM --> Guest Agent (old or [new rust ones](https://gitlab.com/xen-project/xen-guest-agent/-/releases)) --> XCP-ng --> Packer. So without agent at some point the IP cannot be retrieved. 
> while my testing was done go build -o ... and packer plugins install --path ... I did notice that the make generate error happens on master identical to the branch so not sure this should be a blocker.

Let's fix this in another PR 

> @ddelnano if that's the case, what if any hold ups are remaining on this PR?

Time :sweat_smile:

nathanael-h commented:
Oh very nice from you @Erik142 !! Thanks, @gCyrille or I will review this next week I think.

gCyrille commented on builder/xenserver/common/step_wait_for_ip.go:
```diff
@@ -29,47 +29,64 @@ func (self *StepWaitForIP) Run(ctx context.Context, state multistep.StateBag) mu
+					if ip != new_ip {
+						ip = new_ip
+						ui.Message(fmt.Sprintf("Got IP '%s' from HTTP request", ip))
+						state.Put("instance_ssh_address", ip)
```
Don't you want to stop or continue the loop when the IP is found? It will run the next block if the getter is in "auto"

Erik142 replied:
Yes, I do, completely correct! This is an oversight by me, I will update the commit. Thanks!
